test/Pipeline.hs

```diff
@@ -0,0 +1,196 @@
+import qualified Juvix.Core.IR.Typechecker as Typed
+import qualified Juvix.Core.Pipeline as P
+import qualified Juvix.Core.Types as Core
+import qualified Juvix.Core.Usage as Usage
+import Juvix.Library hiding (bool, identity, log)
+import qualified Michelson.Typed as MT
+import qualified Michelson.Untyped as M
+import qualified Test.Tasty as T
+import qualified Test.Tasty.HUnit as T
+import Prelude (String)
+
+data Env primTy primVal
+  = Env
+      { parameterisation :: Core.Parameterisation primTy primVal,
+        log :: [Core.PipelineLog primTy primVal],
+        globals :: IR.Globals primTy primVal
+      }
+  deriving (Generic)
+
+type EnvExecAlias primTy primVal compErr =
+  ExceptT
+    (Core.PipelineError primTy primVal compErr)
+    (StateT (Env primTy primVal) IO)
+
+newtype EnvExec primTy primVal compErr a
+  = EnvE (EnvExecAlias primTy primVal compErr a)
+  deriving (Functor, Applicative, Monad, MonadIO)
+  deriving
+    ( HasSink "log" [Core.PipelineLog primTy primVal],
+      HasWriter "log" [Core.PipelineLog primTy primVal]
+    )
+    via WriterField "log" (EnvExecAlias primTy primVal compErr)
+  deriving
+    ( HasReader "parameterisation" (Core.Parameterisation primTy primVal),
+      HasSource "parameterisation" (Core.Parameterisation primTy primVal)
+    )
+    via ReaderField "parameterisation" (EnvExecAlias primTy primVal compErr)
+  deriving
+    ( HasState "globals" (IR.Globals primTy primVal),
+      HasSource "globals" (IR.Globals primTy primVal),
+      HasSink "globals" (IR.Globals primTy primVal)
+    )
+    via StateField "globals" (EnvExecAlias primTy primVal compErr)
+  deriving
+    (HasReader "globals" (IR.Globals primTy primVal))
+    via ReaderField "globals" (EnvExecAlias primTy primVal compErr)
+  deriving
+    (HasThrow "error" (Core.PipelineError primTy primVal compErr))
+    via MonadError (EnvExecAlias primTy primVal compErr)
+
+exec ::
+  EnvExec primTy primVal CompErr a ->
+  Core.Parameterisation primTy primVal ->
+  IR.Globals primTy primVal ->
+  IO
+    ( Either (Core.PipelineError primTy primVal CompErr) a,
+      [Core.PipelineLog primTy primVal]
+    )
+exec (EnvE env) param globals = do
+  (ret, env) <- runStateT (runExceptT env) (Env param [] globals)
+  pure (ret, log env)
+
+shouldCompileTo ::
+  String ->
+  (HR.Term PrimTy PrimVal, Usage.T, HR.Term PrimTy PrimVal) ->
+  Typed.Globals PrimTy PrimVal ->
+  EmptyInstr ->
+  T.TestTree
+shouldCompileTo name (term, usage, ty) globals instr =
+  T.testCase name $ do
+    res <- toMichelson term usage ty globals
+    show res T.@=? (show (Right instr :: Either String EmptyInstr) :: String)
+
+shouldCompileToContract ::
+  String ->
+  (HR.Term PrimTy PrimVal, Usage.T, HR.Term PrimTy PrimVal) ->
+  Typed.Globals PrimTy PrimVal ->
+  Text ->
+  T.TestTree
+shouldCompileToContract name (term, usage, ty) globals contract =
+  T.testCase name $ do
+    res <- toMichelsonContract term usage ty globals
+    res T.@=? Right contract
```

Comment:
Abstracting out common values and types would be nice

Reply:
Sure, done